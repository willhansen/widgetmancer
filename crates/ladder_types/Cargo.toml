[package]
name = "ladder_types"
version = "0.0.0"
edition.workspace = true

[dependencies]
typenum = "1.17.0"
trybuild = "1.0"
num.workspace = true
derive_more.workspace = true
