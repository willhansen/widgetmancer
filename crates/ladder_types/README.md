A wrapper-type for the phenomenon of subtracting two points resulting in a vector rather than a point.

This crate differs from other approaches (eg euclid) in that rather than the difference of two vectors being a vector, it instead reaches a third level, and so on.

Statically checked

`ThingWithRelativity<Contents, Level=0> ==> T<0>`

`T<0> - T<0> ==> T<1>`
`T<1> - T<1> ==> T<2>`
`T<N> - T<N> ==> T<N+1>`

`T<0> +/- T<1> ==> T<0>`
`T<N> +/- T<N+1> ==> T<0>`

`T<N> + T<M>, where N!=M ==> INVALID`
`T<N> - T<M>, where N!=M and N!=M+1 ==> INVALID`

May have another static parameter for allowing opt-in behavior.  eg allowing `ThingWithRelativity<i32, Level=0>::new(5) + 7`
- Assume no relativity implies level of 0, or whatever the other level is (or one higher?)?
- Maybe an enum for default relativity level?  Options: Same, PlusOne, Zero, Disallow
