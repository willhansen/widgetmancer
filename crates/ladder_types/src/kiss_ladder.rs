use std::ops::{Add, Sub};

#[derive(Eq, PartialEq, Debug)]
pub struct Relative<const LEVEL: u32>();

impl<const L: u32> Relative<L> {
    pub fn new() -> Self {
        Self()
    }
}

impl<const L: u32> Default for Relative<L> {
    fn default() -> Self {
        Self::new()
    }
}

// impl<const LEVEL: u32> Relative<LEVEL> {
//     fn level() -> u32
// }

pub trait LadderRung: Default {}

impl<const L: u32> LadderRung for Relative<L> {}

macro_rules! impl_binary_op {
    ($Trait:ident, $Fn:ident, $Ra:ty, $Rb:ty, $Rc:ty) => {
        impl $Trait<$Rb> for $Ra {
            type Output = $Rc;

            fn $Fn(self, _rhs: $Rb) -> Self::Output {
                <$Rc>::new()
            }
        }
    };
}
macro_rules! impl_add {
    ($Ra:ty, $Rb:ty, $Rc:ty) => {
        impl_binary_op!(Add, add, $Ra, $Rb, $Rc);
    };
}
macro_rules! impl_sub {
    ($Ra:ty, $Rb:ty, $Rc:ty) => {
        impl_binary_op!(Sub, sub, $Ra, $Rb, $Rc);
    };
}
macro_rules! impl_R {
    ($N:literal, $Shorter:ident) => {
        pub type $Shorter = Relative<$N>;
        impl_add!(Relative<{ $N - 1 }>, Relative<$N>, Relative<{ $N - 1 }>);
        impl_sub!(Relative<{ $N - 1 }>, Relative<$N>, Relative<{ $N - 1 }>);
        impl_sub!(Relative<{ $N - 1 }>, Relative<{ $N - 1 }>, Relative<$N>);
    };
}

pub type R0 = Relative<0>;
impl_R!(1, R1);
impl_R!(2, R2);
impl_R!(3, R3);
impl_R!(4, R4);

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_add() {
        assert_eq!(R0::new() + R1::new(), R0::new());
        assert_eq!(R1::new() + R2::new(), R1::new());
    }
    #[test]
    fn test_sub() {
        assert_eq!(R0::new() - R1::new(), R0::new());
        assert_eq!(R1::new() - R2::new(), R1::new());

        assert_eq!(R0::new() - R0::new(), R1::new());
        assert_eq!(R1::new() - R1::new(), R2::new());
    }
}
