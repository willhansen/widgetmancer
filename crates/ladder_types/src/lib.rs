pub mod kiss_ladder;
pub mod typenum_ladder;
pub mod unary_wrapper_ladder;
pub mod numbers_on_ladders;


