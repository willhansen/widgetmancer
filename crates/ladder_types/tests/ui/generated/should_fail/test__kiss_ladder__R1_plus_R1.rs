
use ladder_types::kiss_ladder::*;

fn main() {
   let _ = R1::new() + R1::new();
}
