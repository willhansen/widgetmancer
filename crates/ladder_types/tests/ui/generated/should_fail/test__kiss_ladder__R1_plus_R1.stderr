error[E0308]: mismatched types
 --> tests/ui/generated/should_fail/test__kiss_ladder__R1_plus_R1.rs:5:24
  |
5 |    let _ = R1::new() + R1::new();
  |                        ^^^^^^^^^ expected `2`, found `1`
  |
  = note: expected struct `Relative<2>`
             found struct `Relative<1>`
